diff -Nurp btrfs-progs-v4.4.1.orig/Makefile.in btrfs-progs-v4.4.1/Makefile.in
--- btrfs-progs-v4.4.1.orig/Makefile.in	2016-02-26 19:22:09.000000000 +0200
+++ btrfs-progs-v4.4.1/Makefile.in	2016-03-05 16:00:01.655442036 +0200
@@ -177,7 +177,7 @@ endif
 	@$(check_echo) "    [SP]     $<"
 	$(Q)$(check) $(CFLAGS) $(CHECKER_FLAGS) $<
 	@echo "    [CC]     $@"
-	$(Q)$(CC) $(CFLAGS) -c $<
+	$(Q)$(CC) $(CFLAGS) -fPIC -c $<
 
 %.static.o: %.c
 	@echo "    [CC]     $@"
