--- btrfs-progs-0.19/btrfsck.c.ignore_switch~	2012-03-28 18:05:27.000000000 +0200
+++ btrfs-progs-0.19/btrfsck.c	2012-05-31 18:35:04.518591073 +0200
@@ -3411,6 +3411,9 @@ static struct option long_options[] = {
 	{ "repair", 0, NULL, 0 },
 	{ "init-csum-tree", 0, NULL, 0 },
 	{ "init-extent-tree", 0, NULL, 0 },
+	{ "auto", 0, NULL, 'a' },
+	{ "force", 0, NULL, 'f' },
+	{ "yes", 0, NULL, 'y' },
 	{ 0, 0, 0, 0}
 };
 
@@ -3427,14 +3430,22 @@ int main(int ac, char **av)
 	int option_index = 0;
 	int init_csum_tree = 0;
 	int rw = 0;
+	int ignore = 0;
 
 	while(1) {
 		int c;
-		c = getopt_long(ac, av, "", long_options,
+		c = getopt_long(ac, av, "s:afy", long_options,
 				&option_index);
 		if (c < 0)
 			break;
 		switch(c) {
+			case 'a':
+			case 'f':
+			case 'y':
+				if (!ignore)
+					printf("-a, -f & -y options currently not implemented, ignoring\n");
+				ignore = 1;
+				break;
 			case 's':
 				num = atol(optarg);
 				bytenr = btrfs_sb_offset(num);
