--- btrfs-progs-0.20/mkfs.c~	2013-03-18 18:14:19.000000000 +0100
+++ btrfs-progs-0.20/mkfs.c	2013-03-31 03:52:51.556154829 +0200
@@ -1146,7 +1146,7 @@ static u64 size_sourcedir(char *dir_name
 			allocated_meta_size / default_chunk_size;
 
 	global_total_size = 0;
-	ret = ftw(dir_name, ftw_add_entry_size, 10);
+	ret = nftw(dir_name, ftw_add_entry_size, 10, 0);
 	dir_size = global_total_size;
 	if (ret < 0) {
 		fprintf(stderr, "ftw subdir walk of '%s' failed: %s\n",
