--- btrfs-progs-0.19/Makefile.everything~	2012-05-31 13:13:36.531802228 +0200
+++ btrfs-progs-0.19/Makefile	2012-05-31 13:14:18.197281337 +0200
@@ -20,7 +20,8 @@ RESTORE_LIBS=-lz
 
 progs = btrfsctl mkfs.btrfs btrfs-debug-tree btrfs-show btrfs-vol btrfsck \
 	btrfs btrfs-map-logical btrfs-image btrfs-zero-log btrfs-convert \
-	btrfs-find-root btrfs-restore btrfstune
+	btrfs-find-root btrfs-restore btrfstune btrfs-calc-size \
+	btrfs-corrupt-block btrfs-select-super
 
 # make C=1 to enable sparse
 ifdef C
@@ -43,8 +44,8 @@ btrfs: $(objects) btrfs.o help.o common.
 	$(CC) $(CFLAGS) -o btrfs btrfs.o help.o common.o $(cmds_objects) \
 		$(objects) $(LDFLAGS) $(LIBS) -lpthread
 
-calc-size: $(objects) calc-size.o
-	gcc $(CFLAGS) -o calc-size calc-size.o $(objects) $(LDFLAGS) $(LIBS)
+btrfs-calc-size: $(objects) calc-size.o
+	gcc $(CFLAGS) -o btrfs-calc-size calc-size.o $(objects) $(LDFLAGS) $(LIBS)
 
 btrfs-find-root: $(objects) find-root.o
 	gcc $(CFLAGS) -o btrfs-find-root find-root.o $(objects) $(LDFLAGS) $(LIBS)
