diff -Naur btrfs-progs-v4.8.2/Makefile.in btrfs-progs-v4.8.2.tpg/Makefile.in
--- btrfs-progs-v4.8.2/Makefile.in	2016-10-26 17:43:34.000000000 +0000
+++ btrfs-progs-v4.8.2.tpg/Makefile.in	2016-10-26 21:35:05.653782802 +0000
@@ -251,7 +251,7 @@
 	@$(check_echo) "    [SP]     $<"
 	$(Q)$(check) $(CFLAGS) $(CHECKER_FLAGS) $<
 	@echo "    [CC]     $@"
-	$(Q)$(CC) $(CFLAGS) -c $< -o $@ $($(subst -,_,$(@:%.o=%)-cflags))
+	$(Q)$(CC) $(CFLAGS) -fPIC -c $< -o $@ $($(subst -,_,$(@:%.o=%)-cflags))
 
 %.static.o: %.c
 	@echo "    [CC]     $@"
